### PR TITLE
Fixing various websocket bugs

### PR DESCRIPTION
**Pull request by jplafonta**
Several bugs in different websocket:
- Event callbacks can be called after client has no more reference to the websocket object. Fix is to track client references separately and only call client callbacks if there are remaining client references.

WSPP:
- Deadlock caused by sleep in dtor.  Fixed lifetime management using smart pointers and ensuring wspp_impl object is valid until websocket is disconnected.

WinHttp:
- Connect result was not propagated to caller correctly.
- Calling disconnect after connect is called but not yet completed crashes.

### CODE REVIEW

File: Source/WebSocket/hcwebsocket.h [hunk 1]
```diff
@@ -13,34 +13,45 @@ struct hc_websocket_impl
...
+private:
+    HCWebSocketMessageFunction m_clientMessageFunc;
+    HCWebSocketBinaryMessageFunction m_clientBinaryMessageFunc;
+    HCWebSocketCloseEventFunction m_clientCloseEventFunc;
+    void* m_clientContext;
+
+    std::atomic<int> m_clientRefCount{ 0 };
+    std::atomic<int> m_totalRefCount{ 0 };
+    std::shared_ptr<HC_WEBSOCKET> m_extraRefHolder;
+
```
Comment (jasonsandlin) [inline Source/WebSocket/hcwebsocket.h]: m_ ?

File: Source/WebSocket/Websocketpp/websocketpp_websocket.cpp [hunk 2]
```diff
@@ -70,47 +70,13 @@ struct wspp_websocket_impl : public hc_websocket_impl, public std::enable_shared
         m_hcWebsocketHandle{ hcHandle },
         m_uri(hcHandle->uri)
     {
-        
-        m_state = CREATED;
-        m_numSends = 0;
     }
```
[the remark this answers is not in the document]
Comment (jplafonta) [reply Source/WebSocket/Websocketpp/websocketpp_websocket.cpp]: moved inline with declaration

---
In reply to: [278766617](https://github.com/Microsoft/libHttpClient/pull/403#discussion_r278766617) [](ancestors = 278766617)

File: Source/WebSocket/hcwebsocket.cpp [hunk 3]
```diff
@@ -448,26 +572,28 @@ try
...
     if (closeFunc != nullptr)
     {
-        *closeFunc = websocket->callbackCloseEventFunc;
+        *closeFunc = HC_WEBSOCKET::CloseFunc;
     }
 
     if (context != nullptr)
     {
-        *context = websocket->callbackContext;
+        *context = nullptr;
```
[the remark this answers is not in the document]
Comment (jplafonta) [reply Source/WebSocket/hcwebsocket.cpp]: Yes, same signature.  It seems better to not add a second typedef even though wrappers don't need it, the client callbacks still may

---
In reply to: [278771553](https://github.com/Microsoft/libHttpClient/pull/403#discussion_r278771553) [](ancestors = 278771553)

[the remark this answers is not in the document]
Comment (lubeltra) [reply Source/WebSocket/hcwebsocket.cpp]: The client callbacks definitely need the context pointer, if we're not doubling up on typedefs, keep it

File: Source/WebSocket/hcwebsocket.h [hunk 4]
```diff
@@ -13,34 +13,45 @@ struct hc_websocket_impl
...
     http_internal_string subProtocol;
 
-    HCWebSocketMessageFunction callbackMessageFunc;
-    HCWebSocketBinaryMessageFunction callbackBinaryMessageFunc;
-    HCWebSocketCloseEventFunction callbackCloseEventFunc;
-    void* callbackContext;
-
     std::shared_ptr<hc_websocket_impl> impl;
+private:
+    HCWebSocketMessageFunction m_clientMessageFunc;
```
[the remark this answers is not in the document]
Comment (jplafonta) [reply Source/WebSocket/hcwebsocket.h]: sure

---
In reply to: [278771923](https://github.com/Microsoft/libHttpClient/pull/403#discussion_r278771923) [](ancestors = 278771923)

**Pull request merged**